rusk-abi/src/query/hosted.rs

```diff
@@ -4,50 +4,89 @@
 use dusk_bls12_381_sign::{Signature as BlsSignature, APK};
-use dusk_pki::PublicKey;
+use dusk_pki::{PublicKey, PublicSpendKey};
 use dusk_plonk::prelude::Proof;
 use dusk_schnorr::Signature;
+use piecrust_uplink::{ModuleError, ModuleId};
+
+use bytecheck::CheckBytes;
+use rkyv::{Archive, Deserialize, Serialize};
 
-use crate::{PublicInput, QueryType};
+/// Enum representing all possible payment configurations.
+#[derive(Debug, Clone, Copy, Archive, Deserialize, Serialize)]
+#[archive_attr(derive(CheckBytes))]
+#[repr(C)]
+pub enum PaymentInfo {
+    /// Only transparent notes are accepted.
+    Transparent(Option<PublicSpendKey>),
+    /// Only obfuscated notes are accepted.
+    Obfuscated(Option<PublicSpendKey>),
+    /// Any type of note is accepted.
+    Any(Option<PublicSpendKey>),
+}
 
 /// Compute the blake2b hash of the given bytes, returning the resulting scalar.
 /// The output of the hasher is truncated (last nibble) to fit onto a scalar.
 pub fn hash(bytes: Vec<u8>) -> BlsScalar {
-    let str = QueryType::Hash.as_str();
-    piecrust_uplink::host_query(str, bytes)
+    piecrust_uplink::host_query(Query::HASH, bytes)
 }
 
 /// Compute the poseidon hash of the given scalars
 pub fn poseidon_hash(scalars: Vec<BlsScalar>) -> BlsScalar {
-    let str = QueryType::PoseidonHash.as_str();
-    piecrust_uplink::host_query(str, scalars)
+    piecrust_uplink::host_query(Query::POSEIDON_HASH, scalars)
 }
 
 /// Verify a proof is valid for a given circuit type and public inputs
 pub fn verify_proof(
-    ty: CircuitType,
+    verifier_data: Vec<u8>,
     proof: Proof,
     public_inputs: Vec<PublicInput>,
 ) -> bool {
-    let str = QueryType::VerifyProof.as_str();
-    piecrust_uplink::host_query(str, (ty, proof, public_inputs))
+    piecrust_uplink::host_query(
+        Query::VERIFY_PROOF,
+        (verifier_data, proof, public_inputs),
+    )
 }
 
 /// Verify a schnorr signature is valid for the given public key and message
 pub fn verify_schnorr(msg: BlsScalar, pk: PublicKey, sig: Signature) -> bool {
-    let str = QueryType::VerifySchnorr.as_str();
-    piecrust_uplink::host_query(str, (msg, pk, sig))
+    piecrust_uplink::host_query(Query::VERIFY_SCHNORR, (msg, pk, sig))
 }
 
 /// Verify a BLS signature is valid for the given public key and message
 pub fn verify_bls(msg: Vec<u8>, apk: APK, sig: BlsSignature) -> bool {
-    let str = QueryType::VerifyBls.as_str();
-    piecrust_uplink::host_query(str, (msg, apk, sig))
+    piecrust_uplink::host_query(Query::VERIFY_BLS, (msg, apk, sig))
+}
+
+/// Get the current block height.
+pub fn block_height() -> u64 {
+    piecrust_uplink::host_data(Metadata::BLOCK_HEIGHT)
+}
+
+/// Query a contract for the types of payment it accepts.
+pub fn payment_info(module: ModuleId) -> Result<PaymentInfo, ModuleError> {
+    piecrust_uplink::query(module, "payment_info", ())
+}
+
+/// Define a payment info for the contract.
+#[macro_export]
+macro_rules! payment_info {
+    ($info:expr) => {
```

Comment:
I am not too happy about this macro, could it be eliminated?

Reply:
It can be removed, yes. Do you have a suggestion for what the contract writer should write instead?